From patchwork Mon Feb  4 09:59:42 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10795321
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 45FC04B5F
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:04 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 33C8F2ABA4
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:04 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 26ED02ABB7; Mon,  4 Feb 2019 10:00:04 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C78322ABA4
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:03 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728547AbfBDKAC (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 4 Feb 2019 05:00:02 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:19871 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1727964AbfBDKAC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 4 Feb 2019 05:00:02 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id 3DB2013F89E;
        Mon,  4 Feb 2019 10:59:53 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Andre Przywara <andre.przywara@arm.com>,
        Will Deacon <will.deacon@arm.com>,
        Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool v2 1/3] builtin-run: Fix warning when resolving path
Date: Mon,  4 Feb 2019 10:59:42 +0100
Message-Id: <20190204095944.36950-2-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190204095944.36950-1-aastier@freebox.fr>
References: <20190204095944.36950-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

builtin-run.c: In function ‘kvm_run_write_sandbox_cmd.isra.1’:
builtin-run.c:417:28: error: ‘%s’ directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Werror=format-truncation=]
   snprintf(dst, len, "/host%s", resolved_path);
                            ^~   ~~~~~~~~~~~~~

It's because it understands that len is PATH_MAX, the same as
resolved_path's size. This patch handles the case where the string is
truncated, and fixes the warning.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
Reviewed-by: Andre Przywara <andre.przywara@arm.com>
---
 builtin-run.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/builtin-run.c b/builtin-run.c
index 82e2b2e..463a481 100644
--- a/builtin-run.c
+++ b/builtin-run.c
@@ -414,7 +414,9 @@ static void resolve_program(const char *src, char *dst, size_t len)
 		if (!realpath(src, resolved_path))
 			die("Unable to resolve program %s: %s\n", src, strerror(errno));
 
-		snprintf(dst, len, "/host%s", resolved_path);
+		if (snprintf(dst, len, "/host%s", resolved_path) >= (int)len)
+			die("Pathname too long: %s -> %s\n", src, resolved_path);
+
 	} else
 		strncpy(dst, src, len);
 }

From patchwork Mon Feb  4 09:59:43 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10795323
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id EBF976C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:13 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id DA47729835
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:13 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id CED202ABA4; Mon,  4 Feb 2019 10:00:13 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 5623F29835
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:13 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728646AbfBDKAM (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 4 Feb 2019 05:00:12 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:21510 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1727964AbfBDKAL (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 4 Feb 2019 05:00:11 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id 5056813F7E7;
        Mon,  4 Feb 2019 11:00:01 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Andre Przywara <andre.przywara@arm.com>,
        Will Deacon <will.deacon@arm.com>,
        Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool v2 2/3] virtio: fix warning on strncpy
Date: Mon,  4 Feb 2019 10:59:43 +0100
Message-Id: <20190204095944.36950-3-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190204095944.36950-1-aastier@freebox.fr>
References: <20190204095944.36950-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

virtio/net.c: In function ‘virtio_net__tap_init’:
virtio/net.c:336:47: error: argument to ‘sizeof’ in ‘strncpy’ call is the same expression as the source; did you mean to use the size of the destination? [-Werror=sizeof-pointer-memaccess]
   strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
                                               ^
virtio/net.c:348:47: error: argument to ‘sizeof’ in ‘strncpy’ call is the same expression as the source; did you mean to use the size of the destination? [-Werror=sizeof-pointer-memaccess]
   strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
                                               ^

Fix it by using sizeof of destination instead, even if they're the same
size in this case.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
Reviewed-by: Andre Przywara <andre.przywara@arm.com>
---
 virtio/net.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/virtio/net.c b/virtio/net.c
index a05f4cd..35ff2e9 100644
--- a/virtio/net.c
+++ b/virtio/net.c
@@ -331,7 +331,7 @@ static bool virtio_net__tap_init(struct net_dev *ndev)
 			goto fail;
 	} else if (!skipconf) {
 		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
+		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ifr.ifr_name));
 		sin.sin_addr.s_addr = inet_addr(params->host_ip);
 		memcpy(&(ifr.ifr_addr), &sin, sizeof(ifr.ifr_addr));
 		ifr.ifr_addr.sa_family = AF_INET;
@@ -343,7 +343,7 @@ static bool virtio_net__tap_init(struct net_dev *ndev)
 
 	if (!skipconf) {
 		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
+		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ifr.ifr_name));
 		ioctl(sock, SIOCGIFFLAGS, &ifr);
 		ifr.ifr_flags |= IFF_UP | IFF_RUNNING;
 		if (ioctl(sock, SIOCSIFFLAGS, &ifr) < 0)
@@ -372,7 +372,7 @@ static void virtio_net__tap_exit(struct net_dev *ndev)
 		return;
 
 	sock = socket(AF_INET, SOCK_STREAM, 0);
-	strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
+	strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ifr.ifr_name));
 	ioctl(sock, SIOCGIFFLAGS, &ifr);
 	ifr.ifr_flags &= ~(IFF_UP | IFF_RUNNING);
 	if (ioctl(sock, SIOCGIFFLAGS, &ifr) < 0)

From patchwork Mon Feb  4 09:59:44 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10795325
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 9C00C6C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:20 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8AAD629835
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:20 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 7F4B42ABA4; Mon,  4 Feb 2019 10:00:20 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 1C0A92A939
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon,  4 Feb 2019 10:00:20 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729189AbfBDKAS (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 4 Feb 2019 05:00:18 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:22100 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1728656AbfBDKAS (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 4 Feb 2019 05:00:18 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id 6338113F80A;
        Mon,  4 Feb 2019 11:00:09 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Andre Przywara <andre.przywara@arm.com>,
        Will Deacon <will.deacon@arm.com>,
        Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool v2 3/3] kvmtool: 9p: fix overapping snprintf
Date: Mon,  4 Feb 2019 10:59:44 +0100
Message-Id: <20190204095944.36950-4-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190204095944.36950-1-aastier@freebox.fr>
References: <20190204095944.36950-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

virtio/9p.c: In function ‘virtio_p9_create’:
virtio/9p.c:335:21: error: passing argument 1 to restrict-qualified parameter aliases with argument 4 [-Werror=restrict]
  ret = snprintf(dfid->path, size, "%s/%s", dfid->path, name);
                 ~~~~^~~~~~                 ~~~~~~~~~~

Fix it by allocating a temporary string with dfid->path content instead
of overwriting it in-place, which is limited in glibc snprintf with the
__restrict qualifier.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
Reviewed-by: Andre Przywara <andre.przywara@arm.com>
---
 virtio/9p.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/virtio/9p.c b/virtio/9p.c
index 6bae403..ac70dbc 100644
--- a/virtio/9p.c
+++ b/virtio/9p.c
@@ -322,6 +322,7 @@ static void virtio_p9_create(struct p9_dev *p9dev,
 	struct p9_qid qid;
 	struct p9_fid *dfid;
 	char full_path[PATH_MAX];
+	char *tmp_path;
 	u32 dfid_val, flags, mode, gid;
 
 	virtio_p9_pdu_readf(pdu, "dsddd", &dfid_val,
@@ -332,7 +333,13 @@ static void virtio_p9_create(struct p9_dev *p9dev,
 		goto err_out;
 
 	size = sizeof(dfid->abs_path) - (dfid->path - dfid->abs_path);
-	ret = snprintf(dfid->path, size, "%s/%s", dfid->path, name);
+
+	tmp_path = strdup(dfid->path);
+	if (!tmp_path)
+		goto err_out;
+
+	ret = snprintf(dfid->path, size, "%s/%s", tmp_path, name);
+	free(tmp_path);
 	if (ret >= (int)size) {
 		errno = ENAMETOOLONG;
 		if (size > 0)
